File: ios/battery_monitor/Sources/battery_monitor/BatteryStateStreamHandler.swift
import Flutter
import UIKit

/// Stream handler for battery charging state change notifications.
///
/// Listens to `UIDevice.batteryStateDidChangeNotification` and streams
/// integer codes representing the current charging state to Flutter.
///
/// Emitted codes (mirrored on the Dart side as `ChargingState`):
/// - 0 = unknown
/// - 1 = charging
/// - 2 = discharging (unplugged)
/// - 3 = full
class BatteryStateStreamHandler: NSObject, FlutterStreamHandler {

    /// Event sink used to send battery state changes to Flutter.
    private var eventSink: FlutterEventSink?

    func onListen(
        withArguments arguments: Any?,
        eventSink events: @escaping FlutterEventSink
    ) -> FlutterError? {
        self.eventSink = events

        // Required: UIDevice.batteryState and the notification only
        // deliver values once monitoring is enabled.
        UIDevice.current.isBatteryMonitoringEnabled = true

        NotificationCenter.default.addObserver(
            self,
            selector: #selector(batteryStateDidChange),
            name: UIDevice.batteryStateDidChangeNotification,
            object: nil
        )

        sendBatteryState()

        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        NotificationCenter.default.removeObserver(
            self,
            name: UIDevice.batteryStateDidChangeNotification,
            object: nil
        )

        // Battery monitoring is left enabled because the level handler
        // may still need it.
        eventSink = nil

        return nil
    }

    @objc private func batteryStateDidChange(_ notification: Notification) {
        sendBatteryState()
    }

    /// Maps `UIDevice.current.batteryState` to the integer codes the
    /// Dart side translates back into the `ChargingState` enum.
    private func sendBatteryState() {
        let state = UIDevice.current.batteryState

        let stateCode: Int
        switch state {
        case .charging:
            stateCode = 1
        case .full:
            stateCode = 3
        case .unplugged:
            stateCode = 2
        case .unknown:
            stateCode = 0
        @unknown default:
            stateCode = 0
        }

        eventSink?(stateCode)
    }
}

File: android/src/main/kotlin/com/nllewellyn/battery_status/BatteryStateStreamHandler.kt
package com.nllewellyn.battery_status

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.BatteryManager
import io.flutter.plugin.common.EventChannel

/**
 * Stream handler for battery charging state change notifications.
 *
 * Implements EventChannel.StreamHandler to deliver a stream of charging
 * state codes to Flutter. Registers a BroadcastReceiver for
 * Intent.ACTION_BATTERY_CHANGED.
 *
 * Emits integer codes mapped to the Dart `ChargingState` enum:
 * - 0 = unknown
 * - 1 = charging
 * - 2 = discharging
 * - 3 = full
 * - 4 = connectedNotCharging
 *
 * Platform support: all Android versions; ACTION_BATTERY_CHANGED is a sticky
 * broadcast so the initial state is delivered synchronously on subscription.
 */
class BatteryStateStreamHandler(private val context: Context) : EventChannel.StreamHandler {

    /** Event sink used to send battery state changes to Flutter. */
    private var eventSink: EventChannel.EventSink? = null

    /** Broadcast receiver for battery changes. */
    private var batteryReceiver: BroadcastReceiver? = null

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events

        val filter = IntentFilter(Intent.ACTION_BATTERY_CHANGED)

        val stickyIntent = context.registerReceiver(null, filter)
        stickyIntent?.let { intent ->
            sendBatteryState(intent)
        }

        batteryReceiver = object : BroadcastReceiver() {
            override fun onReceive(context: Context, intent: Intent) {
                if (intent.action == Intent.ACTION_BATTERY_CHANGED) {
                    sendBatteryState(intent)
                }
            }
        }

        context.registerReceiver(batteryReceiver, filter)
    }

    override fun onCancel(arguments: Any?) {
        batteryReceiver?.let {
            context.unregisterReceiver(it)
        }
        batteryReceiver = null
        eventSink = null
    }

    /**
     * Reads EXTRA_STATUS from the battery changed intent and maps it to
     * the integer codes the Dart side translates back into the
     * `ChargingState` enum.
     */
    private fun sendBatteryState(intent: Intent) {
        val status = intent.getIntExtra(
            BatteryManager.EXTRA_STATUS,
            BatteryManager.BATTERY_STATUS_UNKNOWN
        )

        val stateCode = when (status) {
            BatteryManager.BATTERY_STATUS_CHARGING -> 1
            BatteryManager.BATTERY_STATUS_FULL -> 3
            BatteryManager.BATTERY_STATUS_DISCHARGING -> 2
            BatteryManager.BATTERY_STATUS_NOT_CHARGING -> 4
            else -> 0
        }

        eventSink?.success(stateCode)
    }
}
